builtin/v12/migration/top.go

```diff
@@ -95,25 +109,85 @@ func MigrateStateTree(ctx context.Context, store cbor.IpldStore, newManifestCID
 		return cid.Undef, xerrors.Errorf("code cid for miner actor not found in new manifest")
 	}
 
-	mm, err := newMinerMigrator(ctx, store, miner12Cid)
+	minerMigrator, err := newMinerMigrator(ctx, store, miner12Cid, cache)
 	if err != nil {
 		return cid.Undef, xerrors.Errorf("failed to create miner migrator: %w", err)
 	}
-	migrations[miner11Cid] = migration.CachedMigration(cache, *mm)
+	migrations[miner11Cid] = migration.CachedMigration(cache, *minerMigrator)
 
 	if len(migrations)+len(deferredCodeIDs) != len(oldManifestData.Entries) {
 		return cid.Undef, xerrors.Errorf("incomplete migration specification with %d code CIDs, need %d", len(migrations), len(oldManifestData.Entries))
 	}
 
+	// Load the state of the market actor from v11 for migration purposes.
+	oldMarketActor, ok, err := actorsIn.GetActorV5(builtin.StorageMarketActorAddr)
+	if err != nil {
+		return cid.Undef, xerrors.Errorf("failed to get market actor: %w", err)
+	}
+
+	if !ok {
+		return cid.Undef, xerrors.New("didn't find market actor")
+	}
+
+	var oldMarketState market11.State
+	if err := store.Get(ctx, oldMarketActor.Head, &oldMarketState); err != nil {
+		return cid.Undef, xerrors.Errorf("failed to get market actor state: %w", err)
+	}
+
 	actorsOut, err := migration.RunMigration(ctx, cfg, cache, store, log, actorsIn, migrations)
 	if err != nil {
 		return cid.Undef, xerrors.Errorf("failed to run migration: %w", err)
 	}
 
-	outCid, err := actorsOut.Flush()
+	// Retrieve the sector deal IDs from the minerMigrator. These IDs are crucial for the migration of market actor state.
+	sectorDealIDs, err := minerMigrator.sectorDeals.Map.Root()
```

Comment:
This approach relies on HAMT root being deterministic independent of insertion order.  Looking at the code this seems plausible but I want to confirm this is an expected property before relying on this

cc @Stebalien @rvagg

Reply:
After checking in with others and realizing that `RunMigration` relies on this being the case when setting results in `actorsOut` to make the top level actors HAMT I am convinced this is not a problem.

Reply:
:thumbsup: that's one of the fundamental properties that makes us so keen on HAMTs for content addressed data structures. If it ends up not being true then we have a bug!